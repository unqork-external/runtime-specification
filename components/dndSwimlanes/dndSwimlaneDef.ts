import { Default, Description } from '@tsed/schema'

// Ex: 'To Do'
export type DndSwimlaneName = string
export type DndDoubleUnderscore = '__'
export type DndComponentsKey = 'Components'
// Ex: '__todo'
export type DndCardViewTypeKey = `${DndDoubleUnderscore}${Lowercase<string>}`

export type DndItemActionAfterDrop = 'move' | 'copy' | 'disable'
export type DndCardCustomCssType = string
export type DndCardCustomCssClass = string

export type JSONValue = string | number | boolean | JSONObject | JSONArray
export interface JSONObject {
  [x: string]: JSONValue
}
export interface JSONArray extends Array<JSONValue> {}

export type DndSwimlanesDisplayCard = {
  id: string
  showCard: boolean
  disabled: boolean
  nestableKey: DndCardViewTypeKey
  [key: string]: unknown
}
export class DndCardData {
  [key: string]: unknown
}
export type DndSwimlanesCardData = Record<DndSwimlaneName, DndCardData[]>

export type DndCardRowData = string | number | boolean | null | object | Date | DndCard

export type DndCard = {
  [key: string]: DndCardRowData
}

export class DndSwimlaneDef {
  @Description('The id of the Swimlane.')
  id: DndCardViewTypeKey

  @Description('The name of the Swimlane displayed at the top.')
  name: string

  @Description('The Cards that exist in this Swimlane.')
  cards: DndCardData[]

  @Description('The order this Swimlane should appear in.')
  index: number

  @Default([])
  @Description('The Swimlane IDs that Cards can be dragged from and dropped here.')
  allowFrom: DndCardViewTypeKey[] = []
}
