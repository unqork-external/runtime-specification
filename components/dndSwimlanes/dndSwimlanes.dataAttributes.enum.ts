export enum DndSwimlanesDataAttributes {
  itemId = 'itemIdAttribute',
  label = 'itemLabelAttribute',
  layoutType = 'layoutAttribute',
  laneId = 'groupingAttribute',
}
