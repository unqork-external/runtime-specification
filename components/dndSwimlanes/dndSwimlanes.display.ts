import { Default, Description } from '@tsed/schema'

import { Display } from '../../componentComposition/display/component.display'

export class DndSwimlanesDisplay extends Display {
  @Description('If drag & drop is enabled for all Swimlanes or not.')
  @Default(true)
  dndDisabled: boolean = false

  @Description('If we should show the Swimlane count at the top of the Swimlane or not.')
  @Default(true)
  showSwimlaneCount: boolean = true
}
