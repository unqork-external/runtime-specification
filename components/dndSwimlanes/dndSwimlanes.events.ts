import { Optional } from '@tsed/schema'

import type { TargetEventOps } from '../../events/component.eventHandlers'
import { ComponentEventHandlers, EventToOps } from '../../events/component.eventHandlers'
import { OperationsArray } from '../../modules'

export type DndSwimlanesEventsToOps = EventToOps & Record<DndSwimlanesEventTypes.ON_SWIMLANES_MOVE, OperationsArray>

export enum DndSwimlanesEventTypes {
  ON_SWIMLANES_MOVE = 'onSwimlanesMove',
}

export class DndSwimlanesTargetEvents implements TargetEventOps {
  [key: string]: EventToOps

  @Optional()
  root: DndSwimlanesEventsToOps
}

export class DndSwimlanesEvents extends ComponentEventHandlers {
  @Optional()
  declare targets?: DndSwimlanesTargetEvents
}
