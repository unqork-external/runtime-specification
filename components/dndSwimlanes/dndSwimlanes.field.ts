import { Default, Description, Example, Optional, Property } from '@tsed/schema'

import type {
  DndCardCustomCssClass,
  DndCardCustomCssType,
  DndItemActionAfterDrop,
  DndSwimlaneName,
} from './dndSwimlaneDef'
import { trimAll } from '../../../utilities'
import type { BaseComponentDefinition } from '../../baseComponentInterface/base.component.definition'
import { Field } from '../../componentComposition/field/component.field.label'
import { InputLabel } from '../../componentComposition/input/component.input'

export type SwimlaneGroupingAttribute = string
export type CustomType = string
export type DndCardViewByTypeMap = Record<DndSwimlaneName, BaseComponentDefinition[]>

export class DndSwimlanesField extends Field {
  @Property(InputLabel)
  @Description(
    trimAll(`
      Appears above the input field to inform what information is being requested. 
      Labels also help to identify the component on the Module Editor.
    `),
  )
  label: InputLabel = new InputLabel()

  @Description('The message to display in empty Swimlanes.')
  emptySwimlaneMessage: string

  @Description('The attribute from the Card data to use for grouping Swimlanes.')
  @Example('status', 'type')
  @Default('status')
  swimlaneGroupingAttribute: SwimlaneGroupingAttribute = 'status'

  @Description('Which values should be picked from the swimlanes data to be used as the Swimlanes value.')
  valueFormattingAttributes: string[] = []

  @Description('The action for the source Container to take after a Card has been dragged from here.')
  @Default('move')
  sourceItemActionAfterDrop: DndItemActionAfterDrop = 'move'

  @Description('The Boolean attribute from the Card data to use for individually disabling Cards in Swimlanes.')
  @Example('disabled', 'isDraggable')
  @Default('disabled')
  cardDisabledAttribute: string = 'disabled'

  @Description(
    `The 'cardDisabledAttribute' is mapping to 'disabled'
    but if users enter "isMovable" they mean the inverse of "isMovable" === "disabled".`.trim(),
  )
  @Default(false)
  cardDisabledAttributeInvert: boolean = false

  @Description('The attribute from the Card data to use for Card View Type.')
  @Example('status', 'type')
  @Default('status')
  cardViewTypeAttribute: string = 'status'

  @Description('The map of Card View Type attributes to JSON arrays of components to build the Card View.')
  @Default({})
  cardViewByTypeMap: DndCardViewByTypeMap = {}

  @Description('Which attribute to use to apply custom CSS classes to the cards.')
  @Example(
    `
      Ex: 'status', 'type'
      If this is the same as 'swimlaneGroupingAttribute', then Cards can be styled differently
      by the swimlanes they are in. However, if 'cardViewCustomTypeAttribute' is different than 
      'swimlaneGroupingAttribute', then you can have cards in the same swimlane with different styles
      based on some other attribute.
      For a JIRA example, if 'cardViewCustomTypeAttribute' is set to 'epicLink', then cards in the
      'To Do' column could be styled different from each other based on which Epic they've been assigned
      to. 'Epic 1' cards are blue, 'Epic 2' cards are red, etc.
      If 'cardViewCustomTypeAttribute' is the same as 'swimlaneGroupingAttribute', then all cards
      in the 'To Do' swimlane are one color. Cards in 'In Progress' are another color. Etc.
    `.trim(),
  )
  @Default('status')
  cardViewCustomCssTypeAttribute: CustomType = 'status'

  @Description('Custom CSS classes to add based on the CustomTypeAttribute.')
  @Example({ 'To Do': 'to-do-css-class', In_Progress: 'inProgressCssClass' })
  @Default({})
  customCssTypeAttributesMap: Record<DndCardCustomCssType, DndCardCustomCssClass> = {}

  @Optional()
  @Description('CSS classes to add to the Swimlane container.')
  containerCustomClass?: string

  @Optional()
  @Description('CSS classes to add to all items in the Swimlanes.')
  itemCustomClass?: string
}
