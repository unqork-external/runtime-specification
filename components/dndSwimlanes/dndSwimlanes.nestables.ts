import { Property } from '@tsed/schema'

import { NestableDef } from '../../nestables/nestable.definition'
import { Nestable, NestableType } from '../../nestables/nestable.types'

export class DndSwimlanesNestables extends NestableDef {
  @Property()
  rows = new ItemsNestable()
  @Property()
  placeholder = new PlaceholderNestable()
}

export class PlaceholderNestable extends Nestable {
  @Property()
  propertyName: string = 'placeholder'

  @Property()
  type: NestableType = NestableType.MATRIX

  @Property()
  children: string = 'placeholderComponents'
}

export class ItemsNestable extends Nestable {
  @Property()
  propertyName: string = 'rows'

  @Property()
  type: NestableType = NestableType.MATRIX

  @Property()
  children: string = 'components'
}
