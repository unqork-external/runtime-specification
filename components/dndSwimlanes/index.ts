export * from './dndSwimlanes.definition'
export * from './dndSwimlanes.dataAttributes.enum'
export * from './item/dndSwimlanesItem.definition'
export * from './lane/dndSwimlanesLane.definition'
export * from './layouts/dndSwimlanesLayouts.definition'
export * from './layouts/dndSwimlanesLayoutControls.definition'
export * from './layouts/DndSwimlanesLayoutControls.dataAttributes.enum'
