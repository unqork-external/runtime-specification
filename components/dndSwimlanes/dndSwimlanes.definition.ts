import { CollectionOf, Const, Default, Description, DiscriminatorValue, MapOf, Optional, Property } from '@tsed/schema'

import { DndSwimlanesField } from './dndSwimlanes.field'
import { DndSwimlanesNestables } from './dndSwimlanes.nestables'
import { DndSwimlanesOptions } from './dndSwimlanes.options'
import { DndSwimlanesStyling } from './dndSwimlanes.styling'
import { DndSwimlanesTargets } from './dndSwimlanes.targets.enum'
import { DND_SWIMLANES_ITEM_ID, DndSwimlanesItem } from './item/dndSwimlanesItem.definition'
import { DndSwimlanesLane } from './lane/dndSwimlanesLane.definition'
import { DndSwimlanesLayouts } from './layouts/dndSwimlanesLayouts.definition'
import { Examples } from '../../../decorators/schema/examples.decorator'
import { TrimmedDescription } from '../../../decorators/schema/trimmedDescription.decorator'
import { ViewTargets } from '../../../decorators/viewTargets/viewTargets.decorator'
import { BaseComponentDefinition } from '../../base-component-interface'
import { Display } from '../../component-composition'
import type { SignalTargets } from '../../signals'
import { targetedStylingExample } from '../../styling/targeted.styling.example'

@DiscriminatorValue('dndSwimlanes')
@ViewTargets(DndSwimlanesTargets)
export class DndSwimlanesComponentDefinition extends BaseComponentDefinition {
  @Const('dndSwimlanes')
  type: 'dndSwimlanes' = 'dndSwimlanes' as const

  @Optional()
  field: DndSwimlanesField = new DndSwimlanesField()

  @Optional()
  display: Display = new Display()

  @Optional()
  options: DndSwimlanesOptions = new DndSwimlanesOptions()

  @Optional()
  @Default({})
  @MapOf(DndSwimlanesItem)
  @TrimmedDescription(`
    The items that are inside the swimlanes. 
    This can be used to add, remove, or update a swimlane item.
    If the value prop is set, this will be automatically updated reflect the data in value.
  `)
  items: Record<DND_SWIMLANES_ITEM_ID, DndSwimlanesItem> = {}

  @Optional()
  @Default([])
  @CollectionOf(DndSwimlanesLane)
  @Description('The lanes which will contain the swimlanes items')
  lanes: DndSwimlanesLane[] = []

  @Optional()
  layouts: DndSwimlanesLayouts = new DndSwimlanesLayouts()

  @Optional()
  @Default([])
  @CollectionOf(BaseComponentDefinition)
  @Description('A collection of child components to be displayed when the lane is empty.')
  placeholderComponents: BaseComponentDefinition[] = []

  @Property()
  nestables: DndSwimlanesNestables = new DndSwimlanesNestables()

  @Property()
  rows: string[] = []

  @Property()
  placeholder: string[] = []

  @Optional()
  @Default([])
  @TrimmedDescription(`
    A flat array of data attributes for each item in the swimlanes. 
    These are organized from top to bottom and from left to right.
  `)
  value: Record<string, unknown>[] = []

  @Optional()
  declare signals?: SignalTargets<DndSwimlanesTargets>

  @Optional()
  @Examples(targetedStylingExample)
  declare styling?: DndSwimlanesStyling
}
