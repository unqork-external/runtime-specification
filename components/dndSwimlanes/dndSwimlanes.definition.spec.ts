import { DndSwimlanesComponentDefinition } from './dndSwimlanes.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('DndSwimlanes Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(DndSwimlanesComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
    expect(schema.definitions).toMatchSnapshot()
  })
})
