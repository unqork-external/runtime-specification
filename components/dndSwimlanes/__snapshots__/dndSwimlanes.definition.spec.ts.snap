// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DndSwimlanes Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
]
`;

exports[`DndSwimlanes Contract Tests should ensure the contract is clear 2`] = `
{
  "display": {
    "$ref": "#/definitions/Display",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/DndSwimlanesField",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "items": {
    "additionalProperties": {
      "$ref": "#/definitions/DndSwimlanesItem",
    },
    "description": "The items that are inside the swimlanes. This can be used to add, remove, or update a swimlane item. If the value prop is set, this will be automatically updated reflect the data in value.",
    "type": "object",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "lanes": {
    "description": "The lanes which will contain the swimlanes items",
    "items": {
      "$ref": "#/definitions/DndSwimlanesLane",
    },
    "type": "array",
  },
  "layouts": {
    "$ref": "#/definitions/DndSwimlanesLayouts",
  },
  "nestables": {
    "$ref": "#/definitions/DndSwimlanesNestables",
  },
  "options": {
    "$ref": "#/definitions/DndSwimlanesOptions",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "placeholder": {
    "type": "array",
  },
  "placeholderComponents": {
    "description": "A collection of child components to be displayed when the lane is empty.",
    "items": {
      "$ref": "#/definitions/BaseComponentDefinition",
    },
    "type": "array",
  },
  "rows": {
    "type": "array",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/DndSwimlanesStyling",
  },
  "type": {
    "const": "dndSwimlanes",
    "description": "Name of the type of component.",
    "examples": [
      "dndSwimlanes",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "A flat array of data attributes for each item in the swimlanes. These are organized from top to bottom and from left to right.",
    "items": {
      "default": [],
    },
    "type": "array",
  },
}
`;

exports[`DndSwimlanes Contract Tests should ensure the contract is clear 3`] = `
{
  "BaseComponentDefinition": {
    "description": "Base Component Definition is the fundamental interface any component that wishes to leverage the runtime must follow. It is the agreed upon contract between the Runtime and anything that should be recognized as a component.",
    "properties": {
      "executable": {
        "default": true,
        "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
        "type": "boolean",
      },
      "inputs": {
        "items": {
          "$ref": "#/definitions/InputRef",
        },
        "type": "array",
      },
      "key": {
        "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
        "examples": [
          "I am the key",
          "bestKeyEver",
        ],
        "minLength": 1,
        "type": "string",
      },
      "package": {
        "$ref": "#/definitions/PackageModel",
      },
      "signals": {
        "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
        "examples": [
          {
            "args": {
              "inputs": {
                "propertyPath": "value",
                "required": true,
                "targetKey": "firstName",
              },
            },
            "id": "firstNameWatcher",
            "operations": [
              {
                "options": {
                  "logType": "warn",
                  "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
                },
                "type": "PRINTER",
              },
            ],
            "type": "onWatch",
          },
          {
            "id": "textfieldOnBlur",
            "operations": [
              {
                "options": {
                  "logType": "warn",
                  "thingToPrint": "textfield received blur event from DOM",
                },
                "type": "PRINTER",
              },
            ],
            "type": "onBlur",
          },
          "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
        ],
        "type": "object",
      },
      "styling": {
        "$ref": "#/definitions/StylingModel",
      },
      "type": {
        "description": "Name of the type of component.",
        "minLength": 1,
        "type": "string",
      },
      "validation": {
        "$ref": "#/definitions/Validation",
      },
      "value": {
        "description": "Current value of the component",
        "type": [
          "null",
          "integer",
          "number",
          "string",
          "boolean",
          "array",
          "object",
        ],
      },
    },
    "required": [
      "key",
      "type",
    ],
    "type": "object",
  },
  "Display": {
    "description": "Basic UI settings related to how a component may be displayed and interacted with. 'Hidden' corresponds to visibility. 'Interactive' corresponds to the end-user's ability to interact with input components. 'Display as Text' corresponds to the component being rendered with an input element or as text.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "Setting it to \`true\` hides the component from the view, while \`false\` shows it. Regardless of this setting, the component is always visible in the Module Editor. A hidden component can still receive information from other components.",
        "type": "boolean",
      },
      "interactive": {
        "default": true,
        "description": "When \`false\`, it is not possible to interact with or edit the input field. Its appearance is modified (grayed-out by default) to show its state, the HTML 'disabled' attribute is added to the field, and when hovering over it, the cursor also changes to alert of the disabled status. However, the component can still receive information from other components when it's not interactive.",
        "type": "boolean",
      },
      "readOnlyView": {
        "default": false,
        "description": "Defines how the component will be displayed. When \`false\`, the component is displayed as an input field. When \`true\`, the component will not accept any input from the user, and its value will be rendered inline as text and not as an input component. If no value is present, the word 'None' is rendered. If a prefix or suffix is set for the component, they also appear as inline text.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "DndSwimlanesField": {
    "properties": {
      "label": {
        "$ref": "#/definitions/InputLabel",
      },
      "showLaneCount": {
        "default": true,
        "description": "If, in the header, we should show a count of the number of items in the lane.",
        "type": "boolean",
      },
      "showLaneHeader": {
        "default": true,
        "description": "If we should show the lane headers.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "DndSwimlanesItem": {
    "properties": {
      "data": {
        "description": "The data associated with the item. This will be synced with the item's position in the value property. When a card is moved to another lane, the grouping attribute will be updated to reflect the change. If a form field inside the layout changes, the data will reflect the new value of that field.",
        "type": "object",
      },
      "itemId": {
        "description": "Unique id for the item.",
        "minLength": 1,
        "type": "string",
      },
      "label": {
        "description": "An accessible label for the item",
        "type": "string",
      },
      "laneId": {
        "description": "The current lane of the item.",
        "type": "string",
      },
      "layoutControls": {
        "type": "object",
      },
      "layoutId": {
        "description": "The nestable qualifier used to identify this item's layout.",
        "type": "string",
      },
      "layoutType": {
        "default": "default",
        "description": "The layout type used for this card.",
        "type": "object",
      },
    },
    "required": [
      "data",
      "itemId",
      "layoutControls",
    ],
    "type": "object",
  },
  "DndSwimlanesLane": {
    "properties": {
      "allowFrom": {
        "description": "The lanes where you can drag items from and drop here. If not defined, all lanes are allowed.",
        "items": {
          "type": "string",
        },
        "type": "array",
      },
      "items": {
        "description": "List of item ids that are inside this lane",
        "items": {
          "default": [],
          "type": "string",
        },
        "type": "array",
      },
      "laneId": {
        "description": "The id of the Swimlane.",
        "minLength": 1,
        "type": "string",
      },
      "name": {
        "description": "The name of the Swimlane displayed at the top.",
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "laneId",
      "name",
    ],
    "type": "object",
  },
  "DndSwimlanesLayouts": {
    "additionalProperties": {
      "$ref": "#/definitions/BaseComponentDefinition",
    },
    "properties": {
      "default": {
        "description": "The default set of components to render for a swimlane item. Other layouts can be added, but at least the default layout must be set. This will be applied when no other layout fits the item.",
        "items": {
          "$ref": "#/definitions/BaseComponentDefinition",
        },
        "type": "array",
      },
    },
    "required": [
      "default",
    ],
    "type": "object",
  },
  "DndSwimlanesNestables": {
    "additionalProperties": true,
    "properties": {
      "placeholder": {
        "type": "object",
      },
      "rows": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "DndSwimlanesOptions": {
    "properties": {
      "draggableAttribute": {
        "description": "Syncs an item's draggable state to a data attribute. When the value of this attribute is false the item will not be draggable. If not provided, an item will be draggable by default and can be toggled by setting 'items[itemId].layoutControls.draggable'",
        "examples": [
          "isMoveable",
        ],
        "type": "string",
      },
      "draggableAttributeInvert": {
        "default": false,
        "description": "Flips the value of the 'draggableAttribute'. Set this to true if the attribute's value is the inverse of the desired state (ie. 'pinned').",
        "type": "boolean",
      },
      "groupingAttribute": {
        "description": "Syncs an item's lane to a data attribute. This attribute's value will be updated as cards are moved between lanes. If this property is not set, you have to provide a laneId when you are adding an item or it will not show up.",
        "examples": [
          "status",
          "type",
        ],
        "type": "string",
      },
      "hiddenAttribute": {
        "description": "Syncs an item's visibility to a data attribute. When the value of this attribute is true the item not be rendered. If not provided, an item will be visible by default and can be toggled by setting 'items[itemId].layoutControls.hidden'",
        "examples": [
          "isHidden",
          "isVisible",
        ],
        "type": "string",
      },
      "hiddenAttributeInvert": {
        "default": false,
        "description": "Flips the value of the 'hiddenAttribute'. Set this to true if the attribute's value is the inverse of the desired state (ie. 'isVisible').",
        "type": "boolean",
      },
      "interactiveAttribute": {
        "description": "Syncs an item's interactive state to a data attribute. When the value of this attribute is false the item will have a disabled appearance, will not be draggable, and all child components will be non-interactive. If not provided, an item will be interactive by default and can be toggled by setting 'items[itemId].layoutControls.interactive'",
        "examples": [
          "disabled",
          "active",
        ],
        "type": "string",
      },
      "interactiveAttributeInvert": {
        "default": false,
        "description": "Flips the value of the 'interactiveAttribute'. Set this to true if the attribute's value is the inverse of the desired state (ie. 'disabled').",
        "type": "boolean",
      },
      "itemIdAttribute": {
        "description": "The data attribute used as the unique id of the swimlane item.",
        "type": "string",
      },
      "itemKey": {
        "default": "$item",
        "description": "Unique property key that will be available to the scope of the descendants.",
        "type": "string",
      },
      "itemLabelAttribute": {
        "description": "The data attribute used as accessible name of the swimlane item.",
        "examples": [
          "name",
        ],
        "type": "string",
      },
      "layoutAttribute": {
        "description": "Syncs an item's layoutType to a data attribute. The value of this attribute should map to a layout in the layouts property. If not provided or if the attribute's value does not have a corresponding layout, the 'default' layout will be used. If not provided, the item's layoutType can be changed by setting 'items[itemId].layoutType'.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "DndSwimlanesStyling": {
    "type": "object",
  },
  "InputLabel": {
    "description": "Details pertaining to the label element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "When \`true\`, it hides the label of the component. Note the label will still be displayed on the Module Editor.",
        "type": "boolean",
      },
      "text": {
        "description": "Configurable text that is displayed above the input field.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputRef": {
    "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
    "properties": {
      "alias": {
        "type": "string",
      },
      "propertyPath": {
        "type": "string",
      },
      "required": {
        "type": "boolean",
      },
      "silent": {
        "type": "boolean",
      },
      "targetKey": {
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "required": [
      "targetKey",
    ],
    "type": "object",
  },
  "Label": {
    "properties": {
      "text": {
        "description": "Text displayed as the label for the field",
        "type": "string",
      },
    },
    "type": "object",
  },
  "PackageModel": {
    "description": "The PackageModel describes how the Vega Runtime will resolve dependencies",
    "properties": {
      "type": {
        "description": "Type of package",
        "type": "string",
      },
      "url": {
        "description": "URL of package",
        "type": "string",
      },
      "version": {
        "description": "Version of package",
        "type": "string",
      },
    },
    "type": "object",
  },
  "StylingModel": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "Validation": {
    "description": "Holds the validation rules and the most current validation results for the component",
    "properties": {
      "isDirty": {
        "description": "Whether or not field has had user interaction. It will be set to true on user interaction.",
        "type": "boolean",
      },
      "result": {
        "$ref": "#/definitions/ValidationResult",
      },
      "rules": {
        "description": "The set of rules to restrict or define the accepted data for the current component. These are the rules to run for validation of the data input.",
        "examples": [
          "You can use it to define the max length of characters in an input field with a custom error message. rules: { required: { type: ValidationRuleTypes.MAX_LENGTH, validation: { maxLength: 5, }, errorMessage: 'Max length is 5', }, },",
        ],
        "type": "object",
      },
    },
    "type": "object",
  },
  "ValidationError": {
    "properties": {
      "message": {
        "description": "The error message",
        "type": "string",
      },
      "type": {
        "description": "Type of the error",
        "enum": [
          "required",
          "maxLength",
          "minLength",
          "arrayMinLength",
          "arrayMaxLength",
          "pattern",
          "min",
          "max",
          "mask",
          "number",
          "dateinput",
          "isBeforeDate",
          "isAfterDate",
          "customError",
          "isEmail",
        ],
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "type",
    ],
    "type": "object",
  },
  "ValidationResult": {
    "properties": {
      "errors": {
        "items": {
          "$ref": "#/definitions/ValidationError",
        },
        "type": "array",
      },
      "isValid": {
        "type": [
          "null",
          "boolean",
        ],
      },
    },
    "type": "object",
  },
}
`;
